### PR TITLE
Fix issue 103, improve validation

### PR DESCRIPTION
**Pull request by JosefWN**
- `addUriPath` appends a single path to the full path, `uriPath` sets the path as a whole. Since `addUriPath` is a bit less idiomatic to use when setting the full path I have removed that use from the examples (and by mistake in the `CoapClient` as well, the root cause for the bug in #103).
- Setting the `uri` (internal setter) would because of my re-write override the `uriPath` with that of the URI (blank), removed this since it's the base URI that's actually set (host/port).
- Removed some nullables where the function would throw an error if the value was null.
- Made properties symmetric, i.e. `locationPath` setter and getter both return the same thing (almost, sanitized input), rather than setting `locationPath` and getting the same value from `locationPathsString`.
- Removed _nearly_ duplicate properties (`uriPath` vs `uriPathsString`).
- Improved input validation (for example uri path length check only took place on `addUriPath`, but not when setting `uriPath`). Added tests to cover edge cases.

### CODE REVIEW

File: test/coap_message_api_test.dart [hunk 1]
```diff
@@ -242,77 +242,109 @@ void main() {
...
     expect(message.uriPaths.length, 4);
-    expect(message.uriPathsString, 'a/uri/path/longer');
-    expect(() => message.addUriPath(null), throwsArgumentError);
+    expect(message.uriPath, 'a/uri/path/longer');
+    expect(
+      () => message.addUriPath('multiple/not/allowed'),
+      throwsArgumentError,
+    );
+    expect(
+      () => message.addLocationPath('no-double-slash//'),
```
Comment (JosefWN) [inline test/coap_message_api_test.dart:258]: Technically I think an URI path option can be empty, so `a//b/c` might be a valid URI. Not sure how to feel about that, the reason I added this check is I didn't want people adding multiple paths using `addLocationPath` (since it's intended for adding a single path option or path segment).

**Comment by shamblett** [comment]
I wasn't sure if this was ready, so I merged the others, this has now resulted in a conflict.

**Comment by JosefWN** [comment]
Rebased, added a couple more test cases and re-ran all examples except the DTLS ones and they looked OK.

I'm fine with merging this if you guys are.
